### ISSUE TITLE
Shape error combining `custom_jvp` with `vmap`.

### ISSUE THREAD

**Issue by sschoenholz** [google/jax#5849]
Recently I've been using `custom_jvp` with `vmap` and I'm seeing something that looks like a shape error that might involve the multiply operation. The code involves computing the pairwise displacement between points that are parameterized by a position in the unit square, $u_i \in [0,1]^2$, along with an affine transformation $T$ so that $r_i = Tu_i$. We use a `custom_jvp` because we want to compute gradients wrt $r_i$ rather than $u_i$. In traditional JAX fashion, we compute the displacement between two points and then use `vmap` to vectorize. When we compute the gradient of a function wrt the affine transformation we observe (I think) erroneous shapes of the gradient that appear to ignore the vmap.

Here is a repro. Let me know if anything is unclear, if you need more information, or if it turns out I'm just confused.

```python
import jax.numpy as np
from jax.api import vmap, grad, custom_jvp

from jax import lax
from jax import random

from functools import partial

# Transformation Code.

@custom_jvp
def transform(box, R):
  if np.isscalar(box) or box.size == 1:
    return R * box
  elif box.ndim == 2:
    return np.einsum(f'ij,j->i', box, R)
  raise ValueError()

@transform.defjvp
def transform_jvp(primals, tangents):
  box, R = primals
  dbox, dR = tangents
  return (transform(box, R), dR + transform(dbox, R))
  
def periodic_general(box):
  def displacement_fn(Ra, Rb, **kwargs):
    _box = kwargs.get('box', box)
    return transform(_box, Ra - Rb) 

  return displacement_fn

# Repro:

N = 250

# Box can either be specified as a scalar or a 2x2
# Affine transformation.
scalar_box = 1.0
matrix_box = np.eye(2)

displacement = periodic_general(scalar_box)

# Create some randomly positioned points.
key = random.PRNGKey(0)
R = random.uniform(key, (N, 2))

# A simple energy function as the total square of the
# pairwise distances.
# 
# If `box` is a scalar then
# energy_fn :: R -> R
#
# If `box` is a matrix then
# energy_fn :: R^{2\times2} -> R
def energy_fn(box):
  d = partial(displacement, box=box)
  d = vmap(vmap(d, (None, 0)), (0, None))
  return np.sum(d(R, R) ** 2)

# We can verify that the shapes of the primal are correct.
print(energy_fn(scalar_box))
print(energy_fn(matrix_box))

# However, the shape of the gradient is incorrect only for
# the scalar box case.
print(grad(energy_fn)(scalar_box).shape)  # Prints (250, 250, 2)
print(grad(energy_fn)(matrix_box).shape)  # Prints (2, 2)
```

**Comment by mattjj** [comment]
Thanks for the report! Does this look like the same bug as #5832 ?

**Comment by sschoenholz** [comment]
Naively it looks a little different; `custom_vjp` vs `custom_jvp` and #5832 seems to have an explicit error rather than a silent shape error. But I could easily be wrong!

**Comment by mattjj** [comment]
@sschoenholz just taking a look at this now. Running the above code doesn't seem to generate a shape error for me. Does it for you? (Maybe it was fixed in the interim!)

**Comment by mattjj** [comment]
#6470 should fix this!
